src/server/v2.0/handler/project.go

```diff
@@ -63,3 +241,380 @@ func (a *projectAPI) GetLogs(ctx context.Context, params operation.GetLogsParams
+			return errors.BadRequestError(err)
+		}
+	}
+
+	return nil
+}
+
+func (a *projectAPI) populateProperties(ctx context.Context, p *project.Project) error {
+	if secCtx, ok := security.FromContext(ctx); ok {
+		if sc, ok := secCtx.(*local.SecurityContext); ok {
+			roles, err := pro.ListRoles(sc.User(), p.ProjectID)
+			if err != nil {
+				return err
+			}
+			p.RoleList = roles
+			p.Role = highestRole(roles)
+		}
+	}
+
+	total, err := a.repositoryCtl.Count(ctx, q.New(q.KeyWords{"project_id": p.ProjectID}))
+	if err != nil {
+		return err
+	}
+	p.RepoCount = total
+
+	// Populate chart count property
+	if config.WithChartMuseum() {
+		count, err := api.GetChartController().GetCountOfCharts([]string{p.Name})
+		if err != nil {
+			err = errors.Wrap(err, fmt.Sprintf("get chart count of project %d failed", p.ProjectID))
+			return err
+		}
+
+		p.ChartCount = count
+	}
+	return nil
+}
+
+func getProjectQuotaSummary(ctx context.Context, p *project.Project, summary *models.ProjectSummary) {
+	if !config.QuotaPerProjectEnable() {
+		log.Debug("Quota per project disabled")
+		return
+	}
+
+	q, err := quota.Ctl.GetByRef(ctx, quota.ProjectReference, quota.ReferenceID(p.ProjectID))
+	if err != nil {
+		log.Debugf("failed to get quota for project: %d", p.ProjectID)
+		return
+	}
+
+	summary.Quota = &models.ProjectSummaryQuota{}
+	if hard, err := q.GetHard(); err == nil {
+		lib.JSONCopy(&summary.Quota.Hard, hard)
+	}
+	if used, err := q.GetUsed(); err == nil {
+		lib.JSONCopy(&summary.Quota.Used, used)
+	}
+}
+
+func getProjectMemberSummary(ctx context.Context, p *project.Project, summary *models.ProjectSummary) {
+	var wg sync.WaitGroup
+
+	for _, e := range []struct {
+		role  int
+		count *int64
+	}{
+		{common.RoleProjectAdmin, &summary.ProjectAdminCount},
+		{common.RoleMaintainer, &summary.MaintainerCount},
+		{common.RoleDeveloper, &summary.DeveloperCount},
+		{common.RoleGuest, &summary.GuestCount},
+		{common.RoleLimitedGuest, &summary.LimitedGuestCount},
+	} {
+		wg.Add(1)
+		go func(role int, count *int64) {
+			defer wg.Done()
+
+			total, err := pro.GetTotalOfProjectMembers(p.ProjectID, role)
+			if err != nil {
+				log.Debugf("failed to get total of project members of role %d", role)
+				return
+			}
+
+			*count = total
+		}(e.role, e.count)
+	}
+
+	wg.Wait()
+}
+
+func getProjectRegistrySummary(ctx context.Context, p *project.Project, summary *models.ProjectSummary) {
+	if p.RegistryID <= 0 {
+		return
+	}
+
+	registry, err := replication.RegistryMgr.Get(p.RegistryID)
+	if err != nil {
+		log.Warningf("failed to get registry %d: %v", p.RegistryID, err)
+	} else if registry != nil {
+		registry.Credential = nil
+		lib.JSONCopy(&summary.Registry, registry)
+	}
+}
+
+// Returns the highest role in the role list.
+// This func should be removed once we deprecate the "current_user_role_id" in project API
+// A user can have multiple roles and they may not have a strict ranking relationship
+func highestRole(roles []int) int {
```

Comment:
dupliated?

Reply:
Currently, I don't find a good place to save this function, so two functions exist in `src/server/v2.0/handler` and `src/core/api`.

What about putting this function in `pkg/user` or `pkg/project`?

Reply:
Of course, when the search API was moved to `src/server/v2.0/handler`, the `highestRole ` in `src/core/api` can be removed.